GaussianBandpassFilter -> GaussianFilter

TomBugnon opened the pull request:
per https://github.com/SpikeInterface/spikeinterface/issues/2394

This makes freq_min OR freq_max optional
and refactors to GaussianFilter rather than GaussianBandpassFilter.

So there is now an actual lowpass filter in spikeinterface !

I think it is correct but please double check..!
@DradeAW  I didn't check the details of the normalization of the gaussian, are the resulting traces guaranteed to be positive in all cases if the input signal is positive?

Examples (on nice OFF periods)
highpass
![image](https://github.com/SpikeInterface/spikeinterface/assets/14128078/33096c3f-9f93-4b76-9727-a67af75d185e)

lowpass
![image](https://github.com/SpikeInterface/spikeinterface/assets/14128078/3a71093a-0ccf-45fe-ac9a-f18fedd0a825)

bandpass
![image](https://github.com/SpikeInterface/spikeinterface/assets/14128078/135faea2-eb89-47a5-9f6b-1283542c7f2d)

DradeAW commented on src/spikeinterface/preprocessing/filter_gaussian.py:
```diff
@@ -74,10 +87,17 @@ def get_traces(
+        if self.freq_max is not None:
+            pos_factor = self._create_gaussian(traces.shape[0], self.freq_max)
+        else:
+            pos_factor = np.ones((traces.shape[0],))
```
Shouldn't that be `np.zeros`?

TomBugnon replied:
With np.zeros the resulting traces are negative when applying highpass. 
If the idea for high pass is to remove the filtered from the original signal then I think we want (1 - filtered) rather than (0 - filtered), no?

DradeAW replied:
Ah yes you're right my bad,
I guess I shouldn't be reviewing PRs late in the day ^^'

DradeAW commented on src/spikeinterface/preprocessing/tests/test_filter_gaussian.py:
```diff
@@ -21,7 +21,7 @@ def test_filter_gaussian():
 
-    rec_filtered = gaussian_bandpass_filter(recording)
+    rec_filtered = gaussian_filter(recording)
 
```
Maybe we can add tests for lowpass and highpass filters?

TomBugnon replied:
yes.. but I'm not sure what the tests should look like

DradeAW replied:
I guess you can create white noise (each point in time is independent and drawn from a Gaussian), and we can check that after a lowpass/highpass filter, the high/low frequencies are removed (while low/high are unaffected)?

DradeAW reviewed (commented):
Nice work, thank you!

I noted a couple of things :)

DradeAW commented:
> I think it is correct but please double check..! @DradeAW I didn't check the details of the normalization of the gaussian, are the resulting traces guaranteed to be positive in all cases if the input signal is positive?

For the lowpass filter, yes it will always be positive. For the highpass and bandpass it's not guaranteed.

DradeAW commented:
> For the lowpass filter, yes it will always be positive. For the highpass and bandpass it's not guaranteed.

Expanding on that: any highpass filter will lead to negative values,
Simple proof : since you are removing the frequency '0 Hz', then the mean of your signal is going to be 0 (under the hypothesis that the recording is infinite in length).
To have a mean of 0, you need both negative and positive values :)

alejoe91 commented:
Hi @TomBugnon @DradeAW 

What's the status of this? Any additional changes needed on your side?

alejoe91 commented:
@TomBugnon pinging on this ;)

TomBugnon commented:
Hi @alejoe91  @DradeAW 

- I modified current code to use reflect padding because padding with zeros leads to enormous edge effects

- I added some sort of testing having a look at the power density with this type of filtering.

Basically the low and high pass filters (the parts I added) seem to work alright, except the transition bands are huge (and much wider for larger cutoff frequencies), so I had to make the testing really really lenient 
![image](https://github.com/SpikeInterface/spikeinterface/assets/14128078/3710e979-a10d-48f1-92ac-20dd18ff73d0)

When used as a bandpass filter, the traces look kinda alright,
(for instance this is 10-50Hz bandpass, compared with original and butterworth)
![image](https://github.com/SpikeInterface/spikeinterface/assets/14128078/7b2ceb44-f6e1-40a5-bb8b-0a7adc883567)
... But the power spectrum is largely attenuated in the passband (I think due to the wide and variable transition bands)
![image](https://github.com/SpikeInterface/spikeinterface/assets/14128078/f7a5d8d0-be2c-46b8-bf84-37517b2cd0f2)

So I'm not testing that frequency content is conserved in the passband. 

I'm not sure what filter response is expected exactly, but in any case (after the reflect-padding change) the low-pass filtered signal is identical to what you get with the scipy function.

Sorry but I really need to stop working on this now, so I won't be doing the chunking test or go more in depth with this. Happy to send you a notebook or include some futher changes if you like

Cheers!

DradeAW commented:
Hey Tom!

Thanks for the update,

First of all, I don't think I agree with the latest commit (enforcing freq min < freq max).
I mean it's possible to want to apply a high-pass filter with a cutoff higher than the low-pass filter (I don't know why someone would want that, but why remove the possibility?)

Also, the response (from what I'm seeing) looks normal. The Gaussian filter is less steep than a high-order Butterworth/Bessel, but this is the reason why it introduces no overshoot / ringing / rebound.

alejoe91 commented on src/spikeinterface/preprocessing/filter_gaussian.py:
```diff
@@ -7,40 +7,50 @@
 
+        if freq_min is None and freq_max is None:
+            raise ValueError("At least one of `freq_min`,`freq_max` should be specified.")
+        if freq_min is not None and freq_max is not None and freq_max <= freq_min:
+            raise ValueError("Expecting `freq_min`<`freq_max`.")
```
```suggestion
```
I agree with @DradeAW. Actually specifying `freq_min`>`freq_max` allows to use this as a "band-stop" filter

DradeAW commented:
Interesting, I should take a look at that when I have the time

DradeAW commented:
I think I may have made a mistake,

Applying two filters means multiplying them in the Fourier domain, so I believe it should be `traces_fft * pos_factor * (1 - neg_factor)`.
Since usually `freq_min` and `freq_max` are far apart, I didn't see it.

Let me make some check to be absolutely sure, and I'll make a PR to fix this!
